src/spikeinterface/extractors/nwbextractors.py

```diff
@@ -116,57 +116,43 @@ def __init__(
-            import fsspec
-            from fsspec.implementations.cached import CachingFileSystem
-
-            self.stream_cache_path = stream_cache_path if stream_cache_path is not None else get_global_tmp_folder()
-            self.cfs = CachingFileSystem(
-                fs=fsspec.filesystem("http"),
-                cache_storage=str(self.stream_cache_path),
-            )
-            self._file_path = self.cfs.open(str(file_path), "rb")
-            f = h5py.File(self._file_path)
-            self.io = NWBHDF5IO(file=f, mode="r", load_namespaces=True)
-
-        elif stream_mode == "ros3":
-            drivers = h5py.registered_drivers()
-            assertion_msg = "ROS3 support not enbabled, use: install -c conda-forge h5py>=3.2 to enable streaming"
-            assert "ros3" in drivers, assertion_msg
-            self._file_path = str(file_path)
-            self.io = NWBHDF5IO(self._file_path, mode="r", load_namespaces=True, driver="ros3")
-
-        else:
-            self._file_path = str(file_path)
-            self.io = NWBHDF5IO(self._file_path, mode="r", load_namespaces=True)
-
-        self._nwbfile = self.io.read()
-        self._es = get_electrical_series(self._nwbfile, self._electrical_series_name)
-
+        self.file_path = file_path
+        self._nwbfile = self.read_nwb_file(
+            file_path=file_path, stream_mode=stream_mode, stream_cache_path=stream_cache_path
+        )
+        electrical_series = get_electrical_series(self._nwbfile, electrical_series_name)
+        # The indices in the electrode table corresponding to this electrical series
+        electrodes_indices = electrical_series.electrodes.data[:]
+        # The table for all the electrodes in the nwbfile
+        electrodes_table = self._nwbfile.electrodes
+
+        ###################
+        # Extract temporal information TODO: Should be a function
+        ###################
         sampling_frequency = None
-        if hasattr(self._es, "rate"):
-            sampling_frequency = self._es.rate
+        if hasattr(electrical_series, "rate"):
+            sampling_frequency = electrical_series.rate
 
-        if hasattr(self._es, "starting_time"):
-            t_start = self._es.starting_time
+        if hasattr(electrical_series, "starting_time"):
+            t_start = electrical_series.starting_time
         else:
             t_start = None
 
         timestamps = None
-        if hasattr(self._es, "timestamps"):
-            if self._es.timestamps is not None:
-                timestamps = self._es.timestamps
-                t_start = self._es.timestamps[0]
+        if hasattr(electrical_series, "timestamps"):
+            if electrical_series.timestamps is not None:
+                timestamps = electrical_series.timestamps
+                t_start = electrical_series.timestamps[0]
 
-        # if rate is unknown, estimate from timestamps
+        # if rate is unknown, estimate from timestamps TODO: Is this possible?
```

Comment:
Always possible with neuroconv util or nwb inspector util, let me know if you have trouble finding it

Reply:
This comment is confusing because of the previous logica if structure. I will remove it. Meanwhile, one question: The TimeSeries in NWB require either the rate or the timestamps, right? In other words, if they don't have the former, they NEED to have the latter, correct?

Reply:
> This comment is confusing because of the previous logica if structure.

Oh, I just meant for the TODO thing that was added here.

But yeah, `TimeSeries` and all their descendants must have either `starting_tiime` + `rate` (two floats) or just `timestamps` (vector of floats, length matching first axes of `data`). The two are mutually exclusive; I don't know that I've _tried_ to specify both however, maybe it's possible?